gapic/samplegen/samplegen.py

```diff
@@ -0,0 +1,411 @@
+        Loops also have a 'body', which contains statments that may
+        use the variables from the header.
+
+        The body statements are validated recursively.
+
+        Raises:
+            UndefinedVariableReference: If an attempted rvalue base is a previously
+                                        undeclared variable.
+            BadLoop: If a "loop" segments has unexpected keywords
+                     or keyword combinatations.
+
+        """
+        segments = set(body.keys())
+        map_args = {self.MAP_KWORD, self.BODY_KWORD}
+
+        # Even though it's valid python to use a variable outside of the lexical
+        # scope it was defined in,
+        #
+        # i.e.
+        #   for m in molluscs:
+        #     handle(m)
+        #   print("Last mollusc: {}".format(m))
+        #
+        # is allowed, the samplegen spec requires that errors are raised
+        # if strict lexical scoping is violated.
+        previous_defs = set(self.var_defs_)
+
+        if {self.COLL_KWORD, self.VAR_KWORD, self.BODY_KWORD} == segments:
+            collection_name = body[self.COLL_KWORD].split(".")[0]
+            # TODO: Once proto info is being passed in, validate the
+            #       [1:] in the collection name.
+            # TODO: resolve the implicit $resp dilemma
+            # if collection_name.startswith("."):
+            #     collection_name = "$resp" + collection_name
+            if collection_name not in self.var_defs_:
+                raise UndefinedVariableReference("Reference to undefined variable: {}"
+                                                 .format(collection_name))
+
+            var = body[self.VAR_KWORD]
+            self._handle_lvalue(var)
+
+        elif map_args <= segments:
+            segments -= map_args
+            segments -= {self.KEY_KWORD, self.VAL_KWORD}
+            if segments:
+                raise BadLoop("Unexpected keywords in loop statement: {}"
+                              .format(segments))
+
+            map_name_base = body[self.MAP_KWORD].split(".")[0]
+            if map_name_base not in self.var_defs_:
+                raise UndefinedVariableReference("Reference to undefined variable: {}"
+                                                 .format(map_name_base))
+
+            key = body.get(self.KEY_KWORD)
+            if key:
+                self._handle_lvalue(key)
+
+            val = body.get(self.VAL_KWORD)
+            if val:
+                self._handle_lvalue(val)
+
+            if not (key or val):
+                raise BadLoop(
+                    "Need at least one of 'key' or 'value' in a map loop")
+
+        else:
+            raise BadLoop("Unexpected loop form: {}".format(segments))
+
+        self.validate_response(body[self.BODY_KWORD])
+        # Restore the previous lexical scope.
+        # This is stricter than python scope rules
+        # because the samplegen spec mandates it.
+        self.var_defs_ = previous_defs
```

Comment:
You're setting and restoring `var_defs` by hand? Why not have simply pass them as params to the recursive function? It would seem less surprising. (And yes, this would require `_handle_lvalue` to return a new lexical scope [or a delta] and the various clients to either add it to the current one [`define`] or merge&pass it to the recursive function [here]).

Reply:
I'm not convinced that passing and returning (immutable, because why not?)sets would significantly clarify the code, but I'll try it and see.
My original intention was to make an 'activation record' object and store var defs as a stack of activation records, with an iterator that elided the difference between the layers. Then I realized this was total overkill and just went with the copy/restore pattern here.

Reply:
Initial experiment indicates that it's not worth the trouble. The `_validate_define` method needs to return a new scope if it's no longer modifying an instance attribute, which means that the dispatcher needs to catch that scope and thread it into the next statement, which means that all the validate methods need to return a scope, whether they modified the input scope or not. The resulting bookkeeping/overhead/boilerplate obscures clarity.

I'll update the comments on statement validator methods that modify the lexical scope, but unless you feel strongly otherwise I'm going to leave `var_defs_` as a mutable instance attribute.

Reply:
It just smells really bad and error-prone to to do recursion but handle pushing/pulling the state by hand. That's what the stack is for!

is was thinking `updated_defs =  self._handle_lvalue(...)` and then `self.var_defs_ = new_defs` for `define` statements and `self.validate_response(..., updated_defs)` for the recursive call here.

Reply:
The problem there is that define statements can occur in loop bodies, and (to the extent that I understand the restrictions of the lexical scope requirement) these bindings must be invalidated upon exiting the loop body.
I.e. in the follwing,
```
for s in response.squids:
  mantle = s.mantle
  print("Mantle length = {}", mantle.length)
```